components/eamxx/src/physics/mam/impl/mam4_amicphys.cpp

```diff
@@ -1,1202 +1,1486 @@
+
+  // jclea, jcldy   : subarea indices
+  // fclea, fcldy   : area fraction [unitless] of the clear and cloudy subareas
+  // qgcm   (ncnst) : grid cell mean, interstitial constituents (unit does not
+  //                  matter)
+  // qqcwgcm(ncnst) : grid cell mean, cloud-borne  constituents (unit
+  //                  does not matter)
+
+  // qgcmx  (ncnst) : grid cell mean, interstitial constituents (unit does not
+  //                  matter)
+  // qsubx(ncnst,maxsubarea): subarea mixing ratios of interst.
+  //                          constituents(unit does not matter as long as they
+  //                          are consistent with the grid cell mean values)
+
+  // Note: qgcm and qqcwgcm are used for calculating the patitioning factors.
+  // qgcmx is the actual grid cell mean that is partitioned into qsubx.
+
+  for(int imode = 0; imode < num_modes; ++imode) {
+    // calculcate partitioning factors
+
+    // grid cell mean of interstitial aerosol mixing ratio of a single mode
+    Real qgcm_intrst = 0;
+
+    // grid cell mean of cloud-borne  aerosol mixing ratio of a single mode
+    Real qgcm_cldbrn = 0;
+
+    // loop thru all species in a mode
+    for(int ispec = 0; ispec < mam4::num_species_mode(imode); ++ispec) {
+      qgcm_intrst = qgcm_intrst + qgcm[lmassptr_amode_gas_pcnst(ispec, imode)];
+      qgcm_cldbrn =
+          qgcm_cldbrn + qqcwgcm[lmassptr_amode_gas_pcnst(ispec, imode)];
     }
+
+    Real factor_clea;  // partitioning factor for clear  subarea [unitless]
+    Real factor_cldy;  // partitioning factor for cloudy subarea [unitless]
+    get_partition_factors(qgcm_intrst, qgcm_cldbrn, fcldy, fclea,  // in
+                          factor_clea, factor_cldy);               // out
+
+    // apply partitioning factors
+    // Here jclea==1 and jcldy==2
+    for(int ispec = 0; ispec < mam4::num_species_mode(imode); ++ispec) {
+      const int icnst     = lmassptr_amode_gas_pcnst(ispec, imode);
+      qsubx[icnst][jclea] = qgcmx[icnst] * factor_clea;
+      qsubx[icnst][jcldy] = qgcmx[icnst] * factor_cldy;
+    }  // ispec
+  }    // imode
+
+}  // set_subarea_qmass_for_intrst_aerosols
+
+//--------------------------------------------------------------------------------
+//--------------------------------------------------------------------------------
+
+KOKKOS_INLINE_FUNCTION
+void set_subarea_gases_and_aerosols(
+    const int &nsubarea, const int &jclea,                           // in
+    const int &jcldy,                                                // in
+    const Real &fclea, const Real &fcldy,                            // in
+    const Real (&qgcm1)[gas_pcnst], const Real (&qgcm2)[gas_pcnst],  // in
+    const Real (&qqcwgcm2)[gas_pcnst],                               // in
+    const Real (&qgcm3)[gas_pcnst],                                  // in
+    const Real (&qqcwgcm3)[gas_pcnst],                               // in
+    Real (&qsub1)[gas_pcnst][maxsubarea()],                          // out
```

Comment:
Have you considered a typedef for this array type? Something like

```c++
using gas_arr_t = Real[gas_pcnst];
using gas_subarea_arr_t = Real[gas_pcnst][maxsubarea()];
```
and then have the signature of these fcn be
```c++
void set_subarea_gases_and_aerosols(
   ...
  const gas_arr_t qgcm1,  // in
  const gas_arr_t qgcm2,  // in
  ...
  gas_subarea_arr_t qsub1,
  ...
```

Reply:
That's a great idea! We can revise this code.